pydicom/dataset.py

```diff
@@ -2681,6 +2681,30 @@ def __eq__(self, other: object) -> bool:
 
         return NotImplemented
 
+    def __deepcopy__(self, _) -> "FileDataset":
+        """Return a deep copy of the file dataset.
+        Make sure that the filename is not copied in case it is a file-like
+        object.
+
+        Returns
+        -------
+        FileDataset
+            A deep copy of the file data set.
+        """
+        copied = self.__class__(
+            self.filename, self, self.preamble, self.file_meta,
+            self.is_implicit_VR, self.is_little_endian
+        )
+        filename = self.filename
+        self.filename = None
+        for (k, v) in self.__dict__.items():
+            copied.__dict__[k] = copy.deepcopy(v)
+        # if the filename is a string, no copying is needed
+        # if it is a file-like, copying is not wanted, see #1147
+        copied.filename = filename
```

Comment:
Bit simpler:
```python
copied.__dict__ = {k: copy.deepcopy(v) for k, v in self.__dict__.items() if k != 'filename'}
copied.filename = self.filename
```

Otherwise it seems OK to me.

Reply:
Hmm, might be better to set the copy to `None` if not string (with a warning?) since it breaks the deepcopy concept to have a reference to the original object in the copy.

Reply:
Ok, will do this. Probably also completely exclude `filename` from comparisons, what do you think?

Reply:
Hm, I'm really not sure about the `FileDataset.__eq__` operator. Currently there is a test (`test_equality_file_meta`) that checks that file datasets are _not_ equal, if their filenames are different. I see the possibilities:
- leave as is (would compare copied file datasets from file-like as not equal)
- ignore filename if both are not strings (would compare copied file datasets from file-like as equal)
- generally ignore filename in comparison (would only compare the contents)
All have pros and cons, @darcymason, what do you think makes the most sense?

Reply:
I like the third option, because to me I don't think the filename is important - I think two Datasets are equal if all data elements are equal.  In particular, Datasets read from a valid file should have an SOP Instance UID, and two different files should have two different Instance UIDs.

I suppose they should also have the same `is_little_endian` and `is_implicit_VR`, although even that is more of an encoding issue, so I could see even ignoring those.

Unless there is a reason from past discussion to include the filename that I'm forgetting...

Reply:
I agree with that. That was actually what I did implement first (ignoring the file name), but then I noticed that failing test and started to doubt it... If @scaramallion also sees no reason against, I will change this tomorrow. Shall I also exclude the Endianess and VR attributes then?

Reply:
> Shall I also exclude the Endianess and VR attributes then?

To me, I'm about 55/45 on this (slight preference to ignore them), but the easier choice is to keep them, because they have been included in the equality so far, so it is safer to leave that alone in case somehow someone did rely on it.

Reply:
Thank you - I'll leave them alone in this case, provided @scaramallion hasn't a compelling reason to do otherwise.

Reply:
After thinking about it, my preference is that `FileDataset == Dataset` or `FileDataset == FileDataset` should only compare the (non file meta) dataset elements and nothing else.

Reply:
The argument based on FileDataset comparison with Dataset makes a lot of sense to me too.  And it falls on the side of my preference for the endian and implicit/explicit VR.

Reply:
> The argument based on FileDataset comparison with Dataset makes a lot of sense to me too. And it falls on the side of my preference for the endian and implicit/explicit VR.

Ok, I like this, too. Changed the code accordingly, please check!